garden_ai/mlmodel.py

```diff
@@ -215,88 +186,103 @@ def clear_mlflow_staging_directory():
+
+        try:
+            response = requests.get(presigned_download_url, stream=True)
+            response.raise_for_status()
+        except requests.RequestException as re:
+            print(
+                f"Could not download model from presigned url. URL: {presigned_download_url}. Error: {re}"
+            )
+            raise
+
+        try:
+            with open(zip_filepath, "wb") as f:
+                f.write(response.content)
+        except IOError as ioe:
+            print(f"Failed to write model to disk at location {zip_filepath}: {ioe}")
+            raise
+
+        extraction_dir = download_dir / "unzipped"
+        unzipped_path = str(download_dir / extraction_dir)
+
+        try:
+            with zipfile.ZipFile(zip_filepath, "r") as zip_ref:
+                zip_ref.extractall(unzipped_path)
+        except (FileNotFoundError, zipfile.BadZipFile) as fe:
+            print(f"Failed to unzip model directory from Garden model repository. {fe}")
+            raise
+
+        return unzipped_path
+
+    @staticmethod
+    def get_download_url(full_model_name: str) -> str:
+        model_url_json = os.environ.get("GARDEN_MODELS", None)
+        if not model_url_json:
+            raise Exception(
+                "GARDEN_MODELS environment variable was not set. Cannot download model."
+            )
+        try:
+            model_url_dict = json.loads(model_url_json)
+            return model_url_dict[full_model_name]
+        except (json.JSONDecodeError, KeyError):
+            print(
+                f"Could not find url for model {full_model_name} in GARDEN_MODELS env var contents {model_url_json}"
+            )
+            raise
+
+    # Duplicated in this class so that _Model is self-contained
+    def clear_mlflow_staging_directory(self):
+        path = str(self.staging_dir)
+        for item in os.listdir(path):
+            item_path = os.path.join(path, item)
+            if os.path.isfile(item_path):
+                os.remove(item_path)
+            elif os.path.isdir(item_path):
+                shutil.rmtree(item_path)
+
     def _lazy_load_model(self):
         """download and deserialize the underlying model, if necessary."""
         if self.model is None:
-            # don't clutter current directory, especially if running locally
-            self.model = load_model(
-                self.model_uri, suppress_warnings=True, dst_path=MODEL_STAGING_DIR
-            )
+            download_url = self.get_download_url(self.full_name)
+            local_model_path = self.download_and_stage(download_url, self.full_name)
+            self.model = load_model(local_model_path, suppress_warnings=True)
+            try:
+                self.clear_mlflow_staging_directory()
+            except Exception as e:
+                print(
+                    f"Could not clean up model staging directory. Check permissions on {self.staging_dir}"
+                )
+                print(str(e))
+                pass
```

Comment:
I want to keep going here because this isn't a fatal error. There is just some cruft in the staging directory. I'd like to clean it up but the show doesn't have to grind to a halt.